### PR TITLE
Add xapianletor-config and related macros

### PR DESCRIPTION
**Pull request by VaibhavKansagara**

### CODE REVIEW

File: bootstrap [hunk 1]
```diff
@@ -629,7 +629,7 @@ cd ..
 
 # Add -I for in-tree macro directory first, so we use that xapian.m4 in
 # preference to one in any directories which may be added below.
-ACLOCAL="${ACLOCAL-aclocal} -I `pwd`/xapian-core/m4-macros"
+ACLOCAL="${ACLOCAL-aclocal} -I `pwd`/xapian-core/m4-macros -I `pwd`/xapian-letor/m4-macros"
```
Comment (ojwb) [inline bootstrap]: We need to add xapian-core's `m4-macros` so omega, bindings and letor can find `xapian.m4`.  We don't need anything else to find letor's macro (though I guess that might change if bindings for letor are added).

File: bootstrap [hunk 2]
```diff
@@ -898,11 +899,14 @@ for d in $modules ; do
     xapian-core)
       cd "$d" && "$configure" --enable-maintainer-mode --disable-option-checking --cache-file="$here/config.cache" ${1+"$@"}
       ;;
+    xapian-letor)
+      cd "$d" && "$configure" --enable-maintainer-mode --disable-option-checking --cache-file="$here/config.cache" ${1+"$@"}
```
Comment (ojwb) [inline bootstrap]: You need to copy the old `*)` case here - currently `XAPIAN_CONFIG` isn't passed (which is the cause of the build failure you asked about on IRC).

File: bootstrap [hunk 3]
```diff
@@ -898,11 +899,14 @@ for d in $modules ; do
...
+      ;;
     xapian-applications/omega)
       cd "$d" && "$configure" --enable-maintainer-mode --disable-option-checking XAPIAN_CONFIG="$XAPIAN_CONFIG" CPPFLAGS="-I$srcdir/INST/include $user_CPPFLAGS" LDFLAGS="-L$srcdir/INST/lib $user_LDFLAGS" ${1+"$@"}
       ;;
     *)
-      cd "$d" && "$configure" --enable-maintainer-mode --disable-option-checking --cache-file="$here/config.cache" XAPIAN_CONFIG="$XAPIAN_CONFIG" ${1+"$@"}
+      cd "$d" && "$configure" --enable-maintainer-mode --disable-option-checking --cache-file="$here/config.cache" XAPIAN_CONFIG="$XAPIAN_CONFIG" XAPIANLETOR_CONFIG="$XAPIANLETOR_CONFIG" ${1+"$@"}
```
Comment (ojwb) [inline bootstrap]: Again, at least currently nothing needs `XAPIANLETOR_CONFIG` set.

File: bootstrap (hunk 1 above)
Comment (VaibhavKansagara) [reply bootstrap]: @olly I am keeping this change as it is for now as letor bindings are going to be added soon.

Comment (ojwb) [reply bootstrap]: I'm not `@olly`...

But that seems OK to me.

File: xapian-letor/Makefile.am [hunk 4]
```diff
@@ -67,9 +67,34 @@ coverage-check:
 	rm -rf lcov
 	genhtml --demangle-cpp --branch-coverage $(GENHTML_ARGS) -q -p "`pwd`" -t "Test Coverage for xapian-letor `git log -n1 --abbrev-commit --format=%h`" -o lcov xapian-letor.lcov
 
+# We want to clean out build tree paths from the copy of xapianletor-config we
+# install, mostly in the interests of facilitating reproducible builds.
+bin_SCRIPTS = installable/xapianletor-config
+
+# Create the installable version in a subdirectory.  It's not made executable
+# until install time, which hopefully should avoid people trying to use it when
+# building against an uninstalled xapian-core.
```
Comment (ojwb) [inline xapian-letor/Makefile.am]: The comment should say `xapian-letor` here.

File: xapian-letor/cmake/xapianletor-config.cmake.in [hunk 5]
```diff
@@ -0,0 +1,5 @@
+SET(prefix "@prefix@")
+SET(exec_prefix "@exec_prefix@")
+SET(XAPIAN_LIBRARIES "@libdir@/libxapianletor@SHLIBEXT@" CACHE FILEPATH "Libraries for Xapian-letor")
+SET(XAPIAN_INCLUDE_DIR "@incdir@" CACHE PATH "Include path for Xapian-letor")
+SET(XAPIAN_FOUND "TRUE")
```
Comment (ojwb) [inline xapian-letor/cmake/xapianletor-config.cmake.in]: I'm certainly not a cmake expert, but I think we probably need to use different names here to xapian-core (`XAPIANLETOR_LIBRARIES`, etc would seem the obvious choice.)

File: xapian-letor/configure.ac [hunk 6]
```diff
@@ -34,10 +34,30 @@ dnl 0:0:0 1.3.1 First release with xapian-letor split out from xapian-core
 LIBRARY_VERSION_INFO=0:0:0
 AC_SUBST(LIBRARY_VERSION_INFO)
 
+LIBRARY_VERSION_SUFFIX=
+dnl Where xapian.h, etc go.
```
Comment (ojwb) [inline xapian-letor/configure.ac]: `xapian.h` is a header from xapian-core...

File: xapian-letor/configure.ac [hunk 7]
```diff
@@ -34,10 +34,30 @@ dnl 0:0:0 1.3.1 First release with xapian-letor split out from xapian-core
...
+
+    LIBRARY_VERSION_SUFFIX=-$dev_series
+    ;;
+esac
+
+AC_SUBST([LIBRARY_VERSION_SUFFIX])
+
 dnl Where xapian-letor.h, etc go.  In development release append "/xapian-1.3".
 incdir=$includedir
 AC_SUBST([incdir])
```
Comment (ojwb) [inline xapian-letor/configure.ac]: This comment doesn't seem to match the actually behaviour.  Comparing the code with `xapian-core/configure.ac` the code to add there's no equivalent of `incdir=$incdir/xapian-$dev_series` here, which is what implements that behaviour in development releases.

The reason we qualify the include directory paths like this is to ease parallel installation of stable and development versions.  Currently that's not a big concern for letor as there is no stable version.  So either we can make this match core and qualify the path (for future-proofing) or we make the comments match what actually happens here.

File: xapian-letor/configure.ac (hunk 6 above)
Comment (ojwb) [reply xapian-letor/configure.ac]: Also this comment was copied from core, but the line of code it refers to wasn't (it's currently handled further down in letor's `configure.ac` for some reason) so it doesn't actually make sense as the code here currently is.

File: xapian-letor/configure.ac [hunk 8]
```diff
@@ -565,7 +629,19 @@ AC_CONFIG_FILES([
...
 AC_OUTPUT
+
+dnl There are no files generated by AC_OUTPUT in the following directories
+dnl and we need to ensure they exist so that the rest of configure or make
+dnl won't fail because they don't exist when srcdir != builddir.
+if test yes = "$vpath_build" ; then
+  for dir in include/xapian-letor ; do
+    AS_MKDIR_P(["$dir"])
+  done
+fi
```
Comment (ojwb) [inline xapian-letor/configure.ac]: Didn't James already deal with this issue a different way in efaaad5e6eeeff0a85a89da4fe63ca960c149836?

In core we need to create `include/xapian` here because we then go on to generate `include/xapian/version.h`, but there's nothing generated in `include/xapian-letor` at configure-time.

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 9]
```diff
@@ -0,0 +1,172 @@
+# Get XAPIAN_CXXFLAGS, XAPIAN_LIBS, and XAPIAN_VERSION from xapianletor-config and
```
Comment (ojwb) [inline xapian-letor/m4-macros/xapianletor.m4]: This really needs to use a different prefix to xapian-core or else things won't work if the user tries to use both libraries together (and I'd expect almost anything that use xapian-letor would need to use the core API as well).

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+# XAPIANLETOR-CONFIG provides the default name for the xapianletor-config script
+# (which the user can override with "./configure XAPIANLETOR_CONFIG=/path/to/it").
+# If unset, the default is xapianletor-config.  Support for this third parameter
+# was added in Xapian 1.1.3.
+AC_DEFUN([XO_LIB_XAPIANLETOR],
+[
+  AC_ARG_VAR(XAPIANLETOR_CONFIG, [Location of xapianletor-config (default:] ifelse([$3], [], xapianletor-config, [$3]) [on PATH)])
+  dnl AC_PATH_PROG ignores an existing user setting of XAPIANLETOR_CONFIG unless it
+  dnl has a full path, so add special handling for such cases.
+  xapian_config_to_check_for="ifelse([$3], [], xapianletor-config, [$3])"
```
Comment (ojwb) [inline xapian-letor/m4-macros/xapianletor.m4]: I think this is only a local variable so we'll get away with this having the same name as in `xapian.m4`, but it would be less confusing to call this `xapianletor_config_to_check_for` or similar.  Or rename both to something generic like `config_script_to_check_for` and keep the two sources more aligned.

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    ifelse([$2], ,
+      [ifelse([$1], , [
+	dnl Simple check to see if the problem is likely to
+	dnl be that we're using a "packaged" xapian-letor but
+	dnl only have the runtime package installed.
+	for sfx in '' 32 64 ; do
+	  set /usr/lib$sfx/libxapianletor.so.*
+	  if test "/usr/lib$sfx/libxapianletor.so.*" != "$1" ; then
+	    if test -r /etc/debian_version ; then
+	      pkg="libxapian-dev"
```
Comment (ojwb) [inline xapian-letor/m4-macros/xapianletor.m4]: There's no debian package for xapian-letor yet so there isn't a "right answer" here as such, but that's the package name for xapian-core's development libraries and headers, which is definitely a "wrong answer".

File: xapian-letor/Makefile.am (hunk 4 above)
Comment (VaibhavKansagara) [reply xapian-letor/Makefile.am]: fixed in a fixup commit.

File: bootstrap (hunk 2 above)
Comment (VaibhavKansagara) [reply bootstrap]: fixed in a fixup commit.

File: xapian-letor/cmake/xapianletor-config.cmake.in (hunk 5 above)
Comment (VaibhavKansagara) [reply xapian-letor/cmake/xapianletor-config.cmake.in]: fixed in a fixup commit.

File: xapian-letor/configure.ac (hunk 8 above)
Comment (VaibhavKansagara) [reply xapian-letor/configure.ac]: Indeed he did sorry fixed in a fixup commit.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 9 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: fixed in a fixup commit.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 10 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: fixed in a fixup commit.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 11 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: fixed in a fixup commit.

File: xapian-letor/configure.ac (hunk 7 above)
Comment (VaibhavKansagara) [reply xapian-letor/configure.ac]: Here my assumption was that anything which uses letor would eventually also need to use core and hence I am using `LIBRARY_VERSION_SUFFIX` which does not indicate letor version but core version which in turn is substituted in xapianletor-config script as I am adding libxapian-1.5.so and libxapianletor.so to the --ltlibs option in xapianletor-config.in script. Here probably we can change the name to something like `XAPIAN_LIBRARY_VERSION_SUFFIX` as the current name is misleading. Please correct me if I am wrong.

Comment (ojwb) [reply xapian-letor/configure.ac]: When using the `--ltlibs` option libtool will take care of ensuring the appropriate libxapian gets linked in to any executable linked against xapian-letor's library - either the platform handles this automatically (as on Linux for example), or libtool will add libraries listed in the `dependency_libs` entry in `libxapianletor.la` (which is just a text file if you want to take a look at it).

It also seems wrong for xapian-letor to assume what the suffix on libxapian is.  There's no inherent reason why you need to use matching versions of xapian-core and xapian-letor.  I think xapian-letor should really behave just like anything else using the xapian-core API and use `XO_LIB_XAPIAN`/`xapian-config` (or similar) to find out the compiler options to use for linking against xapian-core - these will include any suffix needed.

Comment (VaibhavKansagara) [reply xapian-letor/configure.ac]: understood fixed in a fixup commit.

File: xapian-letor/xapianletor-config.in [hunk 12]
```diff
@@ -313,7 +313,7 @@ for arg in $actions ; do
 	fi
 
 	set_L_to_library_path
-	echo "$F$L -lxapian@LIBRARY_VERSION_SUFFIX@ -lxapianletor"
+	echo "$F$L -lxapian -lxapianletor"
```
Comment (ojwb) [inline xapian-letor/xapianletor-config.in]: We shouldn't need to explicitly link with `-lxapian`.

And doing so will cause problems if the xapian-core that we're using is suffixed, because we actually want `libxapian-1.5` (say) which will get automatically linked in, but then this will cause us to also link to `libxapian` (which will probably be the system install of 1.4.x).  So we end up linking to two different shared libraries which will provide a lot of the same symbols, and things will probably go wrong in a rather messy and confusing way.

**Comment by ojwb** [comment]
The travis failures seem to be due to an extra file left over which 4a27dcf65fcfcc670768dee90396dfca17ed304a should have fixed (the CI ran before I pushed that), so I think they will go away by themselves.

File: xapian-letor/xapianletor-config.in (hunk 12 above)
Comment (VaibhavKansagara) [reply xapian-letor/xapianletor-config.in]: fixed in a fixup commit.

**Comment by VaibhavKansagara** [comment]
@ojwb Is there anything else left to do in this PR?.

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+#
+# If ACTION-IF-FOUND and ACTION-IF-NOT-FOUND are both unset, then an
+# appropriate AC_MSG_ERROR is used as a default ACTION-IF-NOT-FOUND.
+# This allows XO_LIB_XAPIANLETOR to be used without any arguments in the
+# common case where Xapian is a requirement (rather than optional).
+#
+# XAPIANLETOR-CONFIG provides the default name for the xapianletor-config script
+# (which the user can override with "./configure XAPIANLETOR_CONFIG=/path/to/it").
+# If unset, the default is xapianletor-config.  Support for this third parameter
+# was added in Xapian 1.1.3.
```
Comment (jaylett) [inline xapian-letor/m4-macros/xapianletor.m4]: This last sentence isn't true of `XO_LIB_XAPIANLETOR`.

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+# --------------------------------------------------------
+# Check if $XAPIANLETOR_VERSION is at least VERSION.  This macro should
+# be used after XO_LIB_XAPIANLETOR.
+#
+# If ACTION-IF-LESS-THAN is unset, it defaults to an
+# appropriate AC_MSG_ERROR saying that Xapian >= VERSION is needed.
+#
+# If ACTION-IF-GREATHER-THAN-OR-EQUAL is unset, the default is no
+# addtional action.
+AC_DEFUN([XO_REQUIRE],
```
Comment (jaylett) [inline xapian-letor/m4-macros/xapianletor.m4]: Should we really call this just `XO_REQUIRE`? That's the same name as a macro in the `xapian-core` m4 file…

File: xapian-letor/m4-macros/xapianletor.m4 [hunk 15]
```diff
@@ -0,0 +1,150 @@
+# Get XAPIANLETOR_CXXFLAGS, XAPIANLETOR_LIBS, and XAPIANLETOR_VERSION from xapianletor-config and
+# AC_SUBST() them.
+
+# serial 17
```
Comment (jaylett) [inline xapian-letor/m4-macros/xapianletor.m4]: I'm fairly easy on this, but it seems odd to start a new macro file with a serial of something other than 1.

File: xapian-core/m4-macros/xapian-1.5.m4 [hunk 16]
```diff
@@ -31,7 +31,7 @@ AC_DEFUN([XO_LIB_XAPIAN],
   AC_ARG_VAR(XAPIAN_CONFIG, [Location of xapian-config (default:] ifelse([$3], [], xapian-config, [$3]) [on PATH)])
   dnl AC_PATH_PROG ignores an existing user setting of XAPIAN_CONFIG unless it
   dnl has a full path, so add special handling for such cases.
-  xapian_config_to_check_for="ifelse([$3], [], xapian-config, [$3])"
+  config_script_to_check_for="ifelse([$3], [], xapian-config, [$3])"
```
Comment (jaylett) [inline xapian-core/m4-macros/xapian-1.5.m4]: I'm not sure why we're renaming a local variable here? Seems like an unnecessary change, given `XO_LIB_XAPIANLETOR` isn't the same as `XO_LIB_XAPIAN` (these lines in particular will always be different between the two, so it doesn't even make it easier to apply changes from -core to -letor).

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 15 above)
Comment (ojwb) [reply xapian-letor/m4-macros/xapianletor.m4]: I wondered that, but it's also helpful to minimise pointless differences between this and `xapian.m4`.

File: xapian-core/m4-macros/xapian-1.5.m4 (hunk 16 above)
Comment (VaibhavKansagara) [reply xapian-core/m4-macros/xapian-1.5.m4]: >I think this is only a local variable so we'll get away with this having the same name as in xapian.m4, but it would be less confusing to call this xapianletor_config_to_check_for or similar. Or rename both to something generic like config_script_to_check_for and keep the two sources more aligned.
>
@ojwb commented on one of the resolved conversations.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 15 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: Can you explain what is this serial number and it's significance?.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 14 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: fixed in a fixup commit.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 13 above)
Comment (VaibhavKansagara) [reply xapian-letor/m4-macros/xapianletor.m4]: fixed in a fixup commit.

File: xapian-letor/m4-macros/xapianletor.m4 (hunk 15 above)
Comment (jaylett) [reply xapian-letor/m4-macros/xapianletor.m4]: @VaibhavKansagara check out [the autotools documentation on the subject](https://www.gnu.org/software/automake/manual/html_node/Serials.html).

**Comment by ojwb** [comment]
Thanks - rebased, autosquashed and pushed to master as cdce18b8fee7c7ea99978c53c535fdc5c01bd921.

File: xapian-letor/Makefile.am [hunk 17]
```diff
@@ -113,4 +138,4 @@ noinst_HEADERS +=\
 	common/str.h\
 	common/stringutils.h
 
-DISTCHECK_CONFIGURE_FLAGS = "XAPIAN_CONFIG=$(XAPIAN_CONFIG)"
+DISTCHECK_CONFIGURE_FLAGS = "XAPIANLETOR_CONFIG=$(XAPIANLETOR_CONFIG)"
```
Comment (ojwb) [inline xapian-letor/Makefile.am]: I noticed after merging when reviewing #263 (which seems to include a version of the changes here) that this change seems wrong - `xapian-letor` doesn't use `xapianletor-config`, but it ought to use the in-tree xapian-config for the nested build when running `make distcheck`.

Comment (VaibhavKansagara) [reply xapian-letor/Makefile.am]: Can you explain it a bit further as to why it won't use xapianletor-config?.

Comment (VaibhavKansagara) [reply xapian-letor/Makefile.am]: understood.
